### PR TITLE
core/socket: do not assign another fd to SocketPort which already has a fd on deserialization

### LINKED ISSUE #19843: Daemon-reload drops all but one ReusePort sockets
**systemd version the issue has been seen with**
> 248

**Used distribution**
 > Debian

**Linux kernel version used** (`uname -a`)

 > `5.10`

**CPU architecture issue was seen on**
 > AMD64

**Expected behaviour you didn't see**
 > Calling `systemctl daemon-reload` won't affect unchanged socket units.

**Unexpected behaviour you saw**
 > Calling `systemctl daemon-reload` will make an unchanged socket unit drops all but 1 sockets per port.

**Steps to reproduce the problem**
Unit file
```
[Unit]
Description=Test Socket

[Socket]
ListenStream=8081
ListenStream=8081
ListenDatagram=8082
ListenDatagram=8082
ReusePort=True
Service=xxx.service

[Install]
WantedBy=sockets.target
```

After start: 4 sockets as expected
```
ss -tuanlpe | grep 808
udp     UNCONN   0        0                      *:8082                 *:*      users:(("systemd",pid=1,fd=42)) ino:5471866 sk:6001 v6only:0 <->
udp     UNCONN   0        0                      *:8082                 *:*      users:(("systemd",pid=1,fd=43)) ino:5470946 sk:6002 v6only:0 <->
tcp     LISTEN   0        4096                   *:8081                 *:*      users:(("systemd",pid=1,fd=40)) ino:5469175 sk:6003 v6only:0 <->
tcp     LISTEN   0        4096                   *:8081                 *:*      users:(("systemd",pid=1,fd=41)) ino:5476673 sk:6004 v6only:0 <->
```

Next: daemon-reload without changing the unit file
```
systemctl daemon-reload
```

systemctl status shows a warning message:
```
test.socket: Socket unit configuration has changed while unit has been running, and some sock
```

And only 1 socket per port is left.
```
sudo ss -tuanlpe | grep 808
udp     UNCONN   0        0                      *:8082                 *:*      users:(("systemd",pid=1,fd=160)) ino:5470946 sk:6002 v6only:0 <->
tcp     LISTEN   0        4096                   *:8081                 *:*      users:(("systemd",pid=1,fd=158)) ino:5476673 sk:6004 v6only:0 <->
```

### PR DESCRIPTION
**Pull request by yuwata**
Otherwise, if a socket address is duplicated, then the previous fd is closed.

Fixes #19843.

### CODE REVIEW

**Comment by yuwata** [comment]
(I have not tested this.)

**Comment by eaufavor** [comment]
This seems fixes #19843.

**Comment by yuwata** [comment]
@eaufavor Thank you for testing!

File: src/core/socket.c [hunk 1]
```diff
@@ -2689,86 +2682,123 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
                         s->control_command = s->exec_command[id];
                 }
         } else if (streq(key, "fifo")) {
+                bool found = false;
```
Comment (poettering) [inline src/core/socket.c]: move this into the `else` branch?

File: src/core/socket.c [hunk 2]
```diff
@@ -2689,86 +2682,123 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
+                                        p->fd = fdset_remove(fds, fd);
+                                        found = true;
                                         break;
                                 }
+                        if (!found)
+                                log_unit_debug(u, "No matching fifo socket found: %s", value+skip);
+                }
 
         } else if (streq(key, "special")) {
+                bool found = false;
```
Comment (poettering) [inline src/core/socket.c]: ditto

File: src/core/socket.c [hunk 3]
```diff
@@ -2689,86 +2682,123 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
+                                        p->fd = fdset_remove(fds, fd);
+                                        found = true;
                                         break;
                                 }
+                        if (!found)
+                                log_unit_debug(u, "No matching special socket found: %s", value+skip);
+                }
 
         } else if (streq(key, "mqueue")) {
+                bool found = false;
```
Comment (poettering) [inline src/core/socket.c]: ditto

File: src/core/socket.c [hunk 4]
```diff
@@ -2689,86 +2682,123 @@ static int socket_deserialize_item(Unit *u, const char *key, const char *value,
...
+                                        p->fd = fdset_remove(fds, fd);
+                                        found = true;
                                         break;
                                 }
+                        if (!found)
+                                log_unit_debug(u, "No matching mqueue socket found: %s", value+skip);
+                }
 
         } else if (streq(key, "socket")) {
+                bool found = false;
```
Comment (poettering) [inline src/core/socket.c]: ditto

**Comment by poettering** [comment]
lgtm

**Comment by yuwata** [comment]
@poettering Thank you for your review. The above comments are addressed. Upgrading the green label.

**Comment by poettering** [comment]
CI failure seems unrelated.

**Pull request merged**